### ISSUE TITLE
MKL+Cython+OpenMP linear algebra bugs on OSX

### ISSUE THREAD

**Issue by jackkamm** [ContinuumIO/anaconda-issues#8803]
### Actual Behavior

On OSX with MKL-enabled scipy, importing a Cython module that uses OpenMP will cause strange linear algebra behavior. For example, the SVD of a 50x50 identity matrix will have negative values. More specifically, if we do
```
eye = np.eye(51, 51)
u, s, vh = scipy.linalg.svd(eye)
```
Then `u`,`vh`, and `u @ np.diag(s) @ vh` will all contain negative values, as shown here:
![image](https://user-images.githubusercontent.com/9089216/37150697-91640920-22ca-11e8-863f-9135ea2e99a7.png)

However setting `MKL_NUM_THREADS=1` will yield the desired behavior:
![image](https://user-images.githubusercontent.com/9089216/37150735-ba787436-22ca-11e8-89ee-1c6e33b004c3.png)

Note the Cython module only needs to be imported to cause the error -- it does not need to be used in any way.

### Expected Behavior

The singular value decomposition should be correct.

### Steps to Reproduce

Please see https://github.com/jackkamm/osx-mkl-omp-cython-bug for the full details. In brief:

1. Create a clean conda environment. Install python3, cython, gcc, and MKL-enabled numpy/scipy.
2. Create a small Cython module named `ext.pyx` that uses OpenMP:
```
cimport numpy as np
import numpy as np
import cython
from cython.parallel import prange

@cython.boundscheck(False)
@cython.wraparound(False)
@cython.cdivision(True)
def p_arange(int n):
    '''
    A useless function only for demonstration purposes
    '''
    cdef np.ndarray[np.int_t, ndim=1] ret = np.zeros(n)
    for i in range(n, schedule='guided', nogil=True):
        ret[i] = i
    return ret
```
3. Build the Cython module. gcc is required for the `-fopenmp` flag.
4. In Python, import the Cython module, then use scipy to compute the SVD of the identity matrix:
```
import numpy as np
import scipy.linalg

# import the extension but don't do anything with it
import ext

n = 50
eye = np.eye(n+1, n+1)
u, s, vh = scipy.linalg.svd(eye)
eye2 = u @ np.diag(s) @ vh

print("All close: ", np.allclose(eye, eye2)) 
print("Min(eye2): ", np.min(eye2))
```

##### Anaconda or Miniconda version:
Anaconda 3, conda 4.4.11

##### Operating System:

macOS 10.12.6

##### `conda info`

```
     active environment : osx-mkl-omp-cython-bug
    active env location : /Users/jack/anaconda3/envs/osx-mkl-omp-cython-bug
            shell level : 1
       user config file : /Users/jack/.condarc
 populated config files :
          conda version : 4.4.11
    conda-build version : 3.7.0
         python version : 3.6.4.final.0
       base environment : /Users/jack/anaconda3  (writable)
           channel URLs : https://repo.continuum.io/pkgs/main/osx-64
                          https://repo.continuum.io/pkgs/main/noarch
                          https://repo.continuum.io/pkgs/free/osx-64
                          https://repo.continuum.io/pkgs/free/noarch
                          https://repo.continuum.io/pkgs/r/osx-64
                          https://repo.continuum.io/pkgs/r/noarch
                          https://repo.continuum.io/pkgs/pro/osx-64
                          https://repo.continuum.io/pkgs/pro/noarch
          package cache : /Users/jack/anaconda3/pkgs
                          /Users/jack/.conda/pkgs
       envs directories : /Users/jack/anaconda3/envs
                          /Users/jack/.conda/envs
               platform : osx-64
             user-agent : conda/4.4.11 requests/2.18.4 CPython/3.6.4 Darwin/16.7.0 OSX/10.12.6
                UID:GID : 501:20
             netrc file : None
           offline mode : False
```

##### `conda list --show-channel-urls`

```
# packages in environment at /Users/jack/anaconda3/envs/osx-mkl-omp-cython-bug:
#
# Name                    Version                   Build  Channel
ca-certificates           2017.08.26           ha1e5d58_0    defaults
certifi                   2018.1.18                py36_0    defaults
cloog                     0.18.0                        0    defaults
cycler                    0.10.0           py36hfc81398_0    defaults
cython                    0.27.3           py36h6ecb376_0    defaults
freetype                  2.8                  h12048fb_1    defaults
gcc                       4.8.5                         8    defaults
gmp                       6.1.2                hb37e062_1    defaults
intel-openmp              2018.0.0             h8158457_8    defaults
isl                       0.12.2                        1    defaults
kiwisolver                1.0.1            py36h792292d_0    defaults
libcxx                    4.0.1                h579ed51_0    defaults
libcxxabi                 4.0.1                hebd6815_0    defaults
libedit                   3.1                  hb4e282d_0    defaults
libffi                    3.2.1                h475c297_4    defaults
libgfortran               3.0.1                h93005f0_2    defaults
libpng                    1.6.34               he12f830_0    defaults
matplotlib                2.2.0            py36hfa7797c_0    defaults
mkl                       2018.0.1             hfbd8650_4    defaults
mpc                       1.0.3                h7a72875_5    defaults
mpfr                      3.1.5                h711e7fd_2    defaults
ncurses                   6.0                  hd04f020_2    defaults
numpy                     1.14.1           py36ha726252_2    defaults
openssl                   1.0.2n               hdbc3d79_0    defaults
pandas                    0.22.0           py36h0a44026_0    defaults
patsy                     0.5.0                    py36_0    defaults
pip                       9.0.1                    py36_5    defaults
pyparsing                 2.2.0            py36hb281f35_0    defaults
python                    3.6.4                hc167b69_1    defaults
python-dateutil           2.6.1            py36h86d2abb_1    defaults
pytz                      2018.3                   py36_0    defaults
readline                  7.0                  hc1231fa_4    defaults
scipy                     1.0.0            py36h1de22e9_0    defaults
seaborn                   0.8.1            py36h595ecd9_0    defaults
setuptools                38.5.1                   py36_0    defaults
six                       1.11.0           py36h0e22d5e_1    defaults
sqlite                    3.22.0               h3efe00b_0    defaults
statsmodels               0.8.0            py36h9c68fc9_0    defaults
tk                        8.6.7                h35a86e2_3    defaults
tornado                   5.0                      py36_0    defaults
wheel                     0.30.0           py36h5eb2c71_1    defaults
xz                        5.2.3                h0278029_2    defaults
zlib                      1.2.11               hf3cbc9b_2    defaults
```

**Comment by msarahan** [comment]
Yikes.  Fundamentally, what's happening here is that the wrong openmp implementation is being used by MKL.  Your plugin loads gcc's openmp, and MKL can work with that, but it's not detecting it correctly.  Intel is fixing this in future MKL releases, but there is a patch in place that should be working - see https://github.com/AnacondaRecipes/numpy-feedstock/pull/5 and the original report of this error, https://github.com/ContinuumIO/anaconda-issues/issues/7480

We will get in touch with Intel and work this out as soon as we can.  Thanks for reporting, and for the very nice reproduction code.

**Comment by msarahan** [comment]
@oleksandr-pavlyk I have confirmed that that build did use your patch.  Is there something else we need to do here?

**Comment by msarahan** [comment]
In private discussions, @oleksandr-pavlyk suggested that you are probably hitting https://github.com/numpy/numpy/issues/9758, not the issue with OpenMP that I pointed out.  The sinister thing here is that if ANY GUI stuff is loaded before numpy, Mac's Accelerate library gets found preferentially for MKL, which has the strange result that you point to.  Sasha says:

> It would be best if he tried to confirm that by loading NumPy before loading Qt5.  He would be need to do it from a terminal, using Python (not IPython and not jupyter)

**Comment by jackkamm** [comment]
Thanks for your prompt response :)

If I modify the script as follows:
```
import os
import numpy as np
import scipy.linalg

n = 50
eye = np.eye(n+1, n+1)

### do something with scipy; comment to make go BOOM!
scipy.linalg.svd(eye)

# now import the extension
import ext

u, s, vh = scipy.linalg.svd(eye)
eye2 = u @ np.diag(s) @ vh

all_close = np.allclose(eye, eye2)
if all_close:
    print("Whew")
else:
   print("BOOM!")
print("All close?", all_close)
print("Min(eye2):", np.min(eye2))
```
Then the SVD works:
```
Whew
All close? True
Min(eye2): 0.0
```
Commenting the indicated line makes it break:
```
BOOM!
All close? False
Min(eye2): -2.0
```
So the issue seems to arise from scipy not numpy; calling some functions from scipy can make MKL load the right stuff; but merely importing scipy is not enough.

**Comment by oleksandr-pavlyk** [comment]
Thank you @jackkamm, it seems like @msarahan assessment of the problem is likely to be correct. I will investigate and report my findings here.

**Comment by oleksandr-pavlyk** [comment]
After some investigation the issue does boil down to a problem arising from Gnu OpenMP being loaded into memory prior to Intel OpenMP. Math Kernel Library by default uses Intel OpenMP threading layer, and [does not support](https://software.intel.com/en-us/mkl-macos-developer-guide-dynamically-selecting-the-interface-and-threading-layer) GNU threading layer on Mac OS.

Because Gnu OpenMP is already in memory, OS's dynamic loader library resolves OpenMP symbols being requested by Intel MKL threading layer to Gnu OpenMP, if they exist in Gnu OpenMP, otherwise to Intel OpenMP. 

This results in an erratic behavior.

There are several possible work-arounds. 

1. Make sure that script imports numpy, and exercises a function from MKL (such as `np.dot`), prior to loading the native extension to ensure that MKL actually loads `libiomp5.dylib`, prior to loading of the extension.

2. Link native extension against Intel OpenMP. Guided by this [thread](https://software.intel.com/en-us/forums/intel-c-compiler/topic/611485), I modified `setup.py` in your companion repo as follow:

```
extensions = [
        Extension("ext",
                  sources=["ext.pyx"],
                  extra_compile_args=['-fopenmp'],
                  libraries=['iomp5', 'pthread'],
                  include_dirs=[numpy.get_include()])]
```
and was able to build the native extension using your environment with gcc conda-package installed, and had no errors.

3. The best yet, use official C compiler on Mac, which is clang, which comes with XCode, which acts as gcc decoy:

```
[14:01:07 mac02 ~]$ gcc --version
Configured with: --prefix=/Applications/Xcode.app/Contents/Developer/usr --with-gxx-include-dir=/usr/include/c++/4.2.1
Apple LLVM version 7.3.0 (clang-703.0.31)
Target: x86_64-apple-darwin15.4.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```

Using this compiler, you would need to change `"-fopenmp"` switches to compiler and linker to `"-openmp"` switches. 

Doing so will result in linking to LLVM OpenMP, which is compatible with Intel OpenMP, and the code would run as expected.

**Comment by msarahan** [comment]
As a corollary to point 3, you may also use Anaconda's compilers, which are a little more up-to-date from the clang included with MacOS.  There's a blog post at https://www.anaconda.com/blog/developer-blog/utilizing-the-new-compilers-in-anaconda-distribution-5/?lang=en-us and docs at https://conda.io/docs/user-guide/tasks/build-packages/compiler-tools.html

**Comment by mingwandroid** [comment]
The system clang compilers on macOS do not support OpenMP.

This was I've reason we built our own.

The other reasons being C++14 support on macOS 10.9.

Also it seems that you cannot mix and match system libc++ and a custom one and catch exceptions thrown in one in the other.

For these reasons our compilers are a way better idea than macOS system ones.

**Comment by jackkamm** [comment]
Thanks guys -- installing `llvm-openmp` and `clang`, and using the conda `clang` to install, did the trick for me.

Also, the links from @msarahan were super useful, thanks for those. I'm new to conda packaging, and a linux user, and was having a lot of difficulty building my OpenMP-using package for macOS -- those links helped a lot, especially the bit about using
```
build:
  - {{ compiler('c') }}
```
in `meta.yaml`.

**Comment by oleksandr-pavlyk** [comment]
Ready to be resolved?

**Comment by msarahan** [comment]
Validated as fixed in MKL 2018.0.3; test files at https://gist.github.com/msarahan/84403b0e9214aec994326bda2574fc94
